### ISSUE TITLE
Consider adding assert with ENOKI_ASSUME_ALIGNED

### ISSUE THREAD

**Issue by skgbanga** [mitsuba-renderer/enoki#34]
Hello,

First, thanks for making this public. 

While reading the documentation, I noticed this:

```
Performing an aligned load from an unaligned memory address will cause a general protection fault that immediately terminates the application.
```

and correctly, doing an avx512 load on an unaligned memory causes the application to segfault. Would it be possible to add `assert( ptr % n == 0)` whenever you do `ENOKI_ASSUME_ALIGNED`?

e.g. 

```cpp
      static ENOKI_INLINE Derived load_(const void *ptr) {
          return _mm512_load_ps((const Value *) ENOKI_ASSUME_ALIGNED(ptr, 64));
      }
```

to 

```cpp
      static ENOKI_INLINE Derived load_(const void *ptr) {
          assert((uintptr_t) ptr % 64 == 0);
          return _mm512_load_ps((const Value *) ENOKI_ASSUME_ALIGNED(ptr, 64));
      }
```

This catches the problem in debug build.

**Comment by wjakob** [comment]
Hi Sandeep,

Sure-that sounds reasonable. Can you create a PR which adds this for the various backends?

Thanks,
Wenzel

On Mon, Sep 16, 2019, at 7:25 PM, Sandeep Gupta wrote:
> Hello,

> First, thanks for making this public.

> While reading the documentation, I noticed this:

> `Performing an aligned load from an unaligned memory address will cause a general protection fault that immediately terminates the application.
`
> and correctly, doing an avx512 load on an unaligned memory causes the application to segfault. Would it be possible to add `assert( ptr % n == 0)` whenever you do `ENOKI_ASSUME_ALIGNED`?

> e.g.

>       static ENOKI_INLINE Derived load_(const void *ptr) {
          return _mm512_load_ps((const Value *) ENOKI_ASSUME_ALIGNED(ptr, 64));
      }
> to

>       static ENOKI_INLINE Derived load_(const void *ptr) {
          assert((uintptr_t) ptr % 64 == 0);
          return _mm512_load_ps((const Value *) ENOKI_ASSUME_ALIGNED(ptr, 64));
      }

**Comment by skgbanga** [comment]
@wjakob PR #35 created for this. 

Note that my "testing" involved this:

```
cmake -DENOKI_TEST=ON -DCMAKE_BUILD_TYPE=Debug -GNinja ..
ninja
ninja test
```

Additionally, does enoki follow some kind of formatting guide?

**Comment by wjakob** [comment]
Closed via #35.
